### PR TITLE
RFC: Decoupling android specific Doa code from application code

### PR DESCRIPTION
**Pull request by Geigerkind**
Edit: Over the past few days I have started to refactor more and more.

In this PR I want to present a proof of concept on how I plan to decouple the Dao code from the application code. The ultimate goal is to decouple the code enough to build a desktop app on top of.

There are certain things that I want to discuss:
1. I added the pattern language "Repository" to indicate a place to retrieve and store data. I started with the PipeSession. I am open to call it whatever.
2. I split the DB Entity and Application Entity apart to decouple the android specific dao code from the application code. Hence the mapper. This introduces some form of overhead but is essential to establish another persistence layer in the future for the desktop app.
3. I am not to keen on the delete function. Currently it takes a whole entity as an argument. I wondered if an ID would suffice here.
4. Currently the PipeSessionRepository is essentially static, as all functions are in the companion object. I am not sure if this should be the way forward, but seeing that we will go with a strategy pattern, this may be very convenient as we dont have to initialize it somewhere. I dont program much in Kotlin, so maybe you have an elegant solution.
5. Finally I wanted to establish some sort of domain/application core where the application specific code is decoupled from the actual android specific code. This may be as thin as the individual entities, we will see. Probably have to put it into its own module.

### CODE REVIEW

File: app/src/main/kotlin/app/vitune/android/database/mapper/PipedSessionMapper.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package app.vitune.android.database.mapper
```
Comment (dtomvan) [inline app/src/main/kotlin/app/vitune/android/database/mapper/PipedSessionMapper.kt:1]: I don't understand how this makes the current DB system any better. Does it allow for a later module extraction? If so, do you expect someone to come up with a desktop app written in KT?

**Comment by dtomvan** [comment]
Also,


> proof of concept

Do you want this to get merged? Or is this just a WIP reference on your vision for the db code?

File: app/src/main/kotlin/app/vitune/android/database/mapper/PipedSessionMapper.kt (hunk 1 above)
Comment (Geigerkind) [reply app/src/main/kotlin/app/vitune/android/database/mapper/PipedSessionMapper.kt:1]: The current app is written solely for android. So the Database it uses uses an android specific DB that may not be compatible with other database technologies. As the entities and its annotations are also specific to that android db you need a layer of indirection. This is what this is doing. We split the application code from the db specific code and make the application code independent from the platform specific code, enabling us to develop an application with multiple clients or uis if you will so.

**Comment by Geigerkind** [comment]
> Also,
> 
> > proof of concept
> 
> Do you want this to get merged? Or is this just a WIP reference on your vision for the db code?

This is not supposed to be merged yet. I want to show the concept to @25huizengek1 of what I plan to do with the rest of the DB layer. I am currently in the process of doing this for the Song entity.

File: app/src/main/kotlin/app/vitune/android/database/mapper/PipedSessionMapper.kt (hunk 1 above)
Comment (dtomvan) [reply app/src/main/kotlin/app/vitune/android/database/mapper/PipedSessionMapper.kt:1]: I like the enthusiasm here. Thank you for your idea. But that's not what I meant: do you imagine implementing such an app in kotlin or a different programming language?

**Comment by dtomvan** [comment]
Alright, could you mark it as WIP for good measure then?

File: app/src/main/kotlin/app/vitune/android/database/mapper/PipedSessionMapper.kt (hunk 1 above)
Comment (Geigerkind) [reply app/src/main/kotlin/app/vitune/android/database/mapper/PipedSessionMapper.kt:1]: ill probably just use the kotlin composite framework. I dont really fancy building an interface to another programming language.

**Comment by Geigerkind** [comment]
Closing it for now.

**Comment by dtomvan** [comment]
Alright, are you getting stuck or something or is this bukket too big to bite in one PR?

**Comment by Geigerkind** [comment]
I refactored about a third of it, you can still find it in my repo.
I stumbled about other solutions that kind of fit my need for now.
Using InnerTune (mobile) + youtube-music (desktop) currently. Both are also not perfect but they do the bare minimum.

**Comment by dtomvan** [comment]
Alright, thank you. I currently use waydroid with a db I sync maybe every other month for my PC.

**Pull request closed without merging**